internal/states/statefile/version4.go

```diff
@@ -389,12 +427,25 @@ func writeStateV4(file *File, w io.Writer, enc encryption.StateEncryption) tfdia
 				continue
 			}
 
+			hasProviderInstanceKeys := false
+			for _, is := range rs.Instances {
+				if is.ProviderKey != addrs.NoKey {
+					hasProviderInstanceKeys = true
+					break
+				}
+			}
```

Comment:
This is also a little different than where the RFC settled, though in a much less substantial way:

The rule in the RFC was that we write it at the resource level _if all of the instances have the same provider instance address_, regardless of whether that address has a provider instance key.

This distinction is much less significant than the previous one, since it only affects the pretty-unlikely case where an author uses dynamic provider instance keys but nonetheless uses the same provider instance key for every instance of a resource. So in this case it's probably better to just change the RFC, since that situation is unlikely to occur and _even if it did_ the resource-level provider instance key would still not be parseable by older OpenTofu versions anyway.

Reply:
I prefer the option described in the RFC: set resource level provider instance address, if all the instances have the same one, otherwise set instance level provider instance address. It is just simpler and more logical, however, no strong opinion.

Reply:
I believe we've decided to update the RFC rather than the implementation, and so I added an item to https://github.com/opentofu/opentofu/issues/2123 so that we can merge this PR and update the RFC as a separate step later.